File: java-core-plugins/src/main/java/org/unlaxer/tramli/plugins/lint/PolicyLintPlugin.java
package org.unlaxer.tramli.plugins.lint;

import org.unlaxer.tramli.FlowDefinition;
import org.unlaxer.tramli.FlowState;
import org.unlaxer.tramli.plugins.api.AnalysisPlugin;
import org.unlaxer.tramli.plugins.api.PluginDescriptor;
import org.unlaxer.tramli.plugins.api.PluginReport;

import java.util.List;

public final class PolicyLintPlugin<S extends Enum<S> & FlowState> implements AnalysisPlugin<S> {
    private final List<FlowPolicy<S>> policies;

    public PolicyLintPlugin(List<FlowPolicy<S>> policies) {
        this.policies = policies;
    }

    @SuppressWarnings("unchecked")
    public static <S extends Enum<S> & FlowState> PolicyLintPlugin<S> defaults() {
        return new PolicyLintPlugin<>((List<FlowPolicy<S>>) (List<?>) DefaultFlowPolicies.all());
    }

    @Override
    public PluginDescriptor descriptor() {
        return new PluginDescriptor("policy-lint", "Policy Lint", "Applies design-time lint policies to a flow definition.");
    }

    @Override
    public void analyze(FlowDefinition<S> definition, PluginReport report) {
        for (FlowPolicy<S> policy : policies) {
            policy.apply(definition, report);
        }
    }
}

File: ts-plugins/src/lint/policy-lint-plugin.ts
import type { FlowDefinition } from '@unlaxer/tramli';
import type { AnalysisPlugin, PluginDescriptor } from '../api/types.js';
import { PluginReport } from '../api/types.js';
import type { FlowPolicy } from './types.js';
import { allDefaultPolicies } from './default-flow-policies.js';

export class PolicyLintPlugin<S extends string> implements AnalysisPlugin<S> {
  constructor(private readonly policies: FlowPolicy<S>[]) {}

  static defaults<S extends string>(): PolicyLintPlugin<S> {
    return new PolicyLintPlugin(allDefaultPolicies<S>());
  }

  descriptor(): PluginDescriptor {
    return {
      id: 'policy-lint',
      displayName: 'Policy Lint',
      description: 'Applies design-time lint policies to a flow definition.',
    };
  }
  kind() { return 'ANALYSIS' as const; }

  analyze(definition: FlowDefinition<S>, report?: PluginReport): PluginReport {
    const r = report ?? new PluginReport();
    for (const policy of this.policies) {
      policy(definition, r);
    }
    return r;
  }
}
